agent/artifact_uploader.go

```diff
@@ -327,8 +327,10 @@ func (a *ArtifactUploader) upload(ctx context.Context, artifacts []*api.Artifact
 
 				// Update the states of the artifacts in bulk.
 				err := roko.NewRetrier(
-					roko.WithMaxAttempts(10),
-					roko.WithStrategy(roko.Constant(5*time.Second)),
+					// TODO: roko.ExponentialSubsecond(1*time.Second) WithMaxAttempts(10)
+					// Meanwhile, 8 roko.Exponential(2sec) attempts is 1,2,4,8,16,32,64 seconds delay (~2 mins)
+					roko.WithMaxAttempts(8),
+					roko.WithStrategy(roko.Exponential(2*time.Second, 0)),
 				).DoWithContext(ctx, func(r *roko.Retrier) error {
```

Comment:
```
				).DoWithContext(ctx, func(r *roko.Retrier) error {
					timeoutCtx, cancel := context.WithTimeout(ctx, 500 * time.Millisecond)
					defer cancel()

					if _, err := a.apiClient.UpdateArtifacts(timeoutCtx, a.conf.JobID, statesToUpload); err != nil {
```

You should be able to use `context.WithTimeout` to avoid having to set the timeout on the http client. Just make sure you make a new context and only use it in this request attempt!

Reply:
We may want to look into something like this: https://go.dev/blog/http-tracing. Instrumenting the client with [http tracing](https://pkg.go.dev/net/http/httptrace) when in debug mode should give an even better view of where this gets stuck.

Looks like you can also localise it using contexts, so we may want to make it super specific to start with.